### PR TITLE
Eventfd

### PR DESCRIPTION
**Pull request by e-rk**

### CODE REVIEW

File: lib/posix/eventfd.c [hunk 1]
```diff
@@ -49,53 +62,72 @@ static int eventfd_poll_update(struct eventfd *efd,
...
+			z_pend_curr(&efd->lock, key, &efd->wait_q, K_FOREVER);
+		} else {
+			count = (efd->flags & EFD_SEMAPHORE) ? 1 : efd->cnt;
+			efd->cnt -= count;
+			if (efd->cnt == 0)
+				k_poll_signal_reset(&efd->read_sig);
+			break;
+		}
+	}
+	z_unpend_all(&efd->wait_q);
```
Comment (Damian-Nordic) [inline lib/posix/eventfd.c]: probably could be moved to line 106 as it doesn't need to be called in the non-blocking case

File: lib/posix/eventfd.c [hunk 2]
```diff
@@ -223,20 +261,29 @@ int eventfd(unsigned int initval, int flags)
...
-	k_sem_init(&efd->write_sem, 0, UINT32_MAX);
+	efd->cnt = initval;
+
+	k_poll_signal_init(&efd->write_sig);
+	k_poll_signal_init(&efd->read_sig);
+	z_waitq_init(&efd->wait_q);
+
+	if (initval != 0)
+		k_poll_signal_raise(&efd->read_sig, 0);
+	k_poll_signal_raise(&efd->write_sig, 0);
```
Comment (Damian-Nordic) [inline lib/posix/eventfd.c:289]: if (initval < UINT64_MAX - 1)

File: lib/posix/eventfd.c [hunk 3]
```diff
@@ -109,23 +141,34 @@ static ssize_t eventfd_write_op(void *obj, const void *buf, size_t sz)
...
+			z_pend_curr(&efd->lock, key, &efd->wait_q, K_FOREVER);
 		} else {
-			ret = k_sem_take(&efd->write_sem, K_FOREVER);
+			efd->cnt += count;
+			if (efd->cnt == (UINT64_MAX - 1))
+				k_poll_signal_reset(&efd->write_sig);
+			break;
 		}
 	}
+	z_unpend_all(&efd->wait_q);
```
Comment (Damian-Nordic) [inline lib/posix/eventfd.c]: move to line 160?

**Review by Damian-Nordic — APPROVED** [review]
great job!

File: lib/posix/eventfd.c (hunk 2 above)
Comment (e-rk) [reply lib/posix/eventfd.c:289]: Initval is uint32_t, so we don't have to check it. It's some oddity.

**Pull request closed without merging**